grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+    stage::{StageResult, StageStatus},
+    Executor,
+};
+
+use regex::Regex;
+
+use serde::Deserialize;
+
+use tracing::debug;
+
+// default should be longer than run stage default
+const DEFAULT_TIMEOUT: Duration = Duration::from_secs(120);
+
+// exit code used to identify non-signal failures.
+// not configurable via YAML.
+//
+// exit codes >125 are reserved by POSIX; see exit(1p)
+// and https://tldp.org/LDP/abs/html/exitcodes.html
+const ERROR_EXITCODE: i32 = 125;
+
+#[derive(Debug, Default, Deserialize, Clone)]
+pub struct ValgrindConfig {
+    log_file: String,
+    leak_check: Option<bool>,
+    malloc_fill: Option<u8>,
+    free_fill: Option<u8>,
+    suppressions: Option<Vec<String>>,
+}
+
+pub struct Valgrind<E> {
+    executor: E,
+    config: ValgrindConfig,
+    executable: String,
+    args: Vec<String>,
+    stdin: Option<String>,
+    timeout: Option<Duration>,
+}
+
+fn read_file(path: &Path) -> Result<String> {
+    let mut file = File::open(path)?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+    Ok(contents)
+}
+
+impl<E: ProcessExecutor> Valgrind<E> {
+    #[allow(dead_code)]
+    pub fn new(
+        executor: E,
+        config: ValgrindConfig,
+        executable: String,
+        args: Vec<String>,
+        stdin: Option<String>,
+        timeout: Option<Duration>,
+    ) -> Self {
+        Self {
+            executor,
+            config,
+            executable,
+            args,
+            stdin,
+            timeout,
+        }
+    }
+
+    fn gen_cmd(&self, ws: &Path) -> Command {
+        let mut cmd = Command::new("valgrind");
+
+        cmd.add_arg(format!("--log-file={}", &self.config.log_file));
+
+        if let Some(v) = self.config.leak_check {
+            cmd.add_arg(format!("--leak-check={}", if v { "yes" } else { "no" }));
+        }
+
+        cmd.add_arg(format!("--error-exitcode={}", ERROR_EXITCODE));
+
+        if let Some(v) = self.config.malloc_fill {
+            cmd.add_arg(format!("--malloc-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = self.config.free_fill {
+            cmd.add_arg(format!("--free-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                cmd.add_arg(format!("--suppressions={}", supp));
+            }
+        }
+
+        cmd.add_arg("--");
+        cmd.add_arg(&self.executable);
+        cmd.add_args(&self.args);
+        if let Some(v) = &self.stdin {
+            cmd.set_stdin(StdinPipe::Path(v.into()))
+        }
+
+        cmd.set_timeout(self.timeout.unwrap_or(DEFAULT_TIMEOUT));
+        cmd.set_cwd(ws);
+
+        cmd
+    }
+
+    fn read_logfile(&self, ws: &Path) -> Result<String> {
+        let path = ws.join(&self.config.log_file);
+        if !path.exists() {
+            return Err(anyhow!(
+                "Could not find logfile at {:?}. Did Valgrind actually execute?",
+                path.to_str()
```

Comment:
I think[ PathBuf::display](https://doc.rust-lang.org/std/path/struct.PathBuf.html#method.display) is favored for printing out paths.

Reply:
Nice! Will use that instead